### PR TITLE
feat: add page.xpath

### LINKED ISSUE #537: Handle XPath selectors
Is it planned/wanted to implement XPath selectors?

This can be done quite easily, but I don't know if it's a choice to not handle them?

I've implemented two methods in my own code to handle them :

```javascript
async waitForXpath(selector, options = { polling: 'mutation' }) {
  return this.waitForFunction(selector => {
    return null !== document.evaluate(selector, document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue;
  }, options, selector);
}

async $XPath(selector) {
  const remoteObject = await this._rawEvaluate(selector => {
    return document.evaluate(selector, document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue;
  }, selector);

  if (remoteObject.subtype === 'node')
    return new ElementHandle(this._client, remoteObject, this._mouse);

  await utils.releaseObject(this._client, remoteObject);
  return null;
}
```

I can make a PR if you're interested.

### PR DESCRIPTION
**Pull request by JoelEinbinder**
Adds
* page.xpath
* frame.xpath
* elementHandle.xpath

fixes #537

### CODE REVIEW

File: lib/FrameManager.js [hunk 1]
```diff
@@ -283,13 +283,29 @@ class Frame {
...
+    const document = await this._document();
+    const value = await document.$(selector);
+    await document.dispose();
+    return value;
+  }
+
+  /**
+   * @return {!Promise<!ElementHandle>}
+   */
+  async _document() {
```
Comment (aslushnikov) [inline lib/FrameManager.js:298]: don't you want to cache this? Otherwise, everything becomes two-hops instead of a one hop.

**Comment by paulirish** [comment]
I'm not an XPATH user so my opinion matters little, but the inclusion of `elementHandle.xpath()` stands out to me. Might want to see that element-rooted xpath is desired before committing to it in the API. It wasn't mentioned in #537 so it's hard to say..

edit: i made a poll: https://github.com/GoogleChrome/puppeteer/issues/537#issuecomment-352605428

**Comment by JoelEinbinder** [comment]
@paulirish I wanted to move all the query selector code to elementHandle anyway. It comes mostly for free. And `elementHandle.xpath('..')` to get the parent feels pretty nice.

File: lib/FrameManager.js [hunk 2]
```diff
@@ -408,7 +423,7 @@ class Frame {
       const url = options.url;
       try {
         const context = await this._contextPromise;
-        return await context.evaluateHandle(addScriptUrl, url);
+        return (await context.evaluateHandle(addScriptUrl, url)).asElement();
```
Comment (JoelEinbinder) [inline lib/FrameManager.js:426]: I added jsdoc for _contextPromise and lots of errors showed up.

File: lib/FrameManager.js (hunk 1 above)
Comment (JoelEinbinder) [reply lib/FrameManager.js:298]: Done.

**Review by aslushnikov — APPROVED** [review]
Thank you!

**Comment by brikou** [comment]
I think we should also have `page.$x` as we have `page.$` and `page.$$`, WDYT?

**Comment by JoelEinbinder** [comment]
@brikou just as an alias or with different functionality? I think things could quickly get very cryptic.

**Comment by brikou** [comment]
I don't think it is cryptic... current `page.$$`, and `page.$` exists and are meant to work as in browser (`$$`, `$`), the same should go for `$x` (and so `page.$x`).

image

**Comment by JoelEinbinder** [comment]
I didn't realize we had `$x` in DevTools. Learn something new every day.

**Comment by aslushnikov** [comment]
wow, that nice. I'd keep this aligned with devtools command-line api.

@JoelEinbinder wdyt about renaming `page.xpath` into `page.$x`?

**Pull request merged**